README.md
[![CircleCI](https://circleci.com/gh/apostrophecms/apostrophe-headless/tree/master.svg?style=svg)](https://circleci.com/gh/apostrophecms/apostrophe-headless/tree/master)

## Apostrophe as a headless CMS

[Apostrophe](http://apostrophecms.org) is great for building websites, but many projects these days just need a "headless" CMS: an easy way to create new content types by defining schemas and immediately have a friendly interface for managing them on the back end... and REST APIs on the front end for React, React Native and other frontend frameworks to talk to.

Just as often, projects call for a mix of the two: Apostrophe as a CMS for the pages of the site, with React-style apps "mixed in" on certain pages.

The `apostrophe-headless` module provides REST APIs for content types created with Apostrophe's [pieces](https://docs.apostrophecms.org/core-concepts/reusable-content-pieces/reusable-content-with-pieces.html) feature. With this module, you might choose to just click "Page Settings" and lock down the "home page" of your site to "logged in users only," then use Apostrophe as a pure headless CMS... or you might mix and match. It's up to you.

> We'll start out by talking about pieces, because they map so well to REST concepts. But `apostrophe-headless` also supports working with pages. We recommend you read about pieces first to figure out the basics, especially authentication.

## Adding a REST API for products

Let's assume you have a module called `products` that extends `apostrophe-pieces` as described in our [reusable content with pieces](https://docs.apostrophecms.org/core-concepts/reusable-content-pieces/reusable-content-with-pieces.html) tutorial. Now you want a REST API so your app can easily get information about pieces.

## Install the package

```
npm install apostrophe-headless
```

## Turn it on

```javascript
// in app.js
modules: {

  'apostrophe-headless': {},

  'products': {
    // Usually you'll put most of this in lib/products/index.js
    extend: 'apostrophe-pieces',
    name: 'product',
    // etc...
    restApi: true
  }
}
```

### Configuration options

You can also pass options for the REST API:

```javascript
  'products': {
    // etc...
    restApi: {
      // max 50 pieces per API result (the default)
      maxPerPage: 50,
      // Allow the public API to invoke additional
      // cursor filters. Note that most schema
      // fields have a cursor filter available
      safeFilters: [ 'slug' ],
      // Restrict GET routes to users with correct permission (false by default)
      getRequiresEditPermission: true
    }
  }
}
```

> Setting `maxPerPage` high can have performance impacts. Consider designing your app with pagination or infinite scroll in mind rather than fetching thousands of pieces the user will not actually look at.

> All of the documentation below discusses the `products` example above. Of course you may also configure the `restApi` option for other modules that extend pieces.

## Workflow support

When used in the presence of the `apostrophe-workflow`, this module currently only supports GET requests. When making GET requests, you must specify the locale name of interest using the `_workflowLocale` query parameter, otherwise you will get results for the default locale only.

Note that this is not an issue if the document type in question is excluded from workflow via `excludeTypes`.


## Retrieving all the products

Now your app can access:

`/api/v1/products`

To get the first page of products (50 per page, unless `maxPerPage` is adjusted as shown above). The response is JSON. See the `results` property for an array of products included in the first page, and the `pages` property for the total number of pages.

If you want to fetch a second page of products:

`/api/v1/products?page=2`

To avoid performance issues we do not send more than 50 products per API call. Your app should make additional queries as needed.

### Filtering products

Here are some examples:

`/api/v1/products?search=cheese`

`/api/v1/products?autocomplete=che`

There's much more. You can use any [s filter](https://docs.apostrophecms.org/advanced-topics/database/cursors.html) that offers a `sanitize` method via the query string. It's [not hard to add custom filters](https://docs.apostrophecms.org/advanced-topics/database/cursors.html#custom-filters) if you need to, but keep in mind that most schema field types have built-in [filter support](https://docs.apostrophecms.org/advanced-topics/database/cursors.html#built-in-filters-every-schema-field-gets-one).

To call most filters from the public API, you will need to use the `safeFilters` option to declare these filters "safe.". Rather than just `restApi: true`, write:

```javascript
'my-module': {
  restApi: {
    // We're assuming here that you have added fields
    // called 'color' and 'brand' in your schema
    safeFilters: [ 'slug', 'color', 'brand' ]
  }
}
```

> You may filter `joinByOne` and `joinByArray` fields, as long as they are listed in `safeFilters`. When doing so pass the `_id` property. Alternatively, leave the leading `_` off the field name and pass the `slug` property.

### Filtering fields for all requests

You can restrict what fields to send by adding `api: false` to a specific schema field. If only users with editing permissions for the doc should see a specific field, you can pass the option `api: 'editPermissionRequired'`.

```javascript
{
  name: 'specificField',
  label: 'Specific Field',
  type: 'string',
  api: false
}
```

### Filtering fields for a single request

You can also require only specific fields in the GET request by adding the query filters `includeFields` and `excludeFields`.

Examples:

`/api/v1/products?includeFields=type,slug,name`

```
// response example
[
  {
    _id: 'whatever_id',
    type: 'product',
    slug: 'product-key-product'
  }
]
```

The response will contain only `_id`, `type`, `slug` and (if present) `name`.

`/api/v1/products?excludeFields=type,slug,name`

```
// response example
[
  {
    _id: 'whatever_id',
    title: 'Product',
    body: { ... }
  }
]
```

The response will contain everything *except* `type`, `slug` and `name`.

> It is useless to use both `includeFields` and `excludeFields` in the same query, as `includeFields` has priority over `excludeFields`. This is due to the way MongoDB projections work.

If there is any conflict between the `api` schema field option above and the `includeFields` option, the `api` schema field option takes priority.

## Retrieving distinct tags, joins, etc. to populate your filters

In addition to fetching actual pieces, you can obtain information about the distinct tags that may exist on those pieces, as well as information about the distinct objects that are joined to them.

This is useful to populate your filters. For instance, to allow the user to filter the results by tag **without prefetching every result in the database to scan for tags**, you must know what tags exist.

To add information about distinct tags to the response, first configure your module to allow it:

```javascript
// in lib/modules/products/index.js
'products': {
  extend: 'apostrophe-pieces',
  name: 'product',
  restApi: {
    safeDistinct: [ 'tags' ]
  }
}
```

> Without `safeDistinct`, developers would be able to cause a denial of service by requesting all distinct values at once for fields like `_id` that are always different.

Now, you may access URLs like this:

`/api/v1/products?distinct=tags`

The response will look like:

```javascript
{
  results: [ ... pieces here ],
  distinct: {
    tags: [
      {
        label: 'Free',
        value: 'Free'
      },
      {
        label: 'Paid',
        value: 'Paid'
      }
    ]
  }
}
```

Now we can display the labels to our users, and if they pick one, send back the value in the `tags` query parameter:

`/api/v1/products?tags=Paid`

**Since the distinct values are intended for use as filters, use of `safeDistinct` implies `safeFilter` as well.** You don't have to specify both for the same filter.

> You can pass multiple values for `tags`, with or without the familiar `[]`, syntax, for example: `tags[]=one&tags[]=two`
> You'll get results that include at least one of the tags.

### Distinct values for joins

Now let's assume there is a `joinByOne` schema field called `_specialist` that joins our `product` piece with a `specialist` piece. We can fetch distinct values here too. In this case, the `value` property will be the `_id`:

```javascript
// in lib/modules/products/index.js

'products': {
  name: 'product',
  extend: 'apostrophe-pieces',
  addFields: [
    {
      type: 'joinByOne',
      name: '_specialist'
    }
  ],
  restApi: {
    safeDistinct: [ '_specialist' ]
  }
}
```

Then we can access:

`/api/v1/products?distinct=_specialist`

The response will look like:

```javascript
{
  results: [ ... pieces here ],
  distinct: {
    _specialist: [
      {
        label: 'Jane Doe',
        value: '_cyyyy'
      },
      {
        label: 'Joe Smith',
        value: '_czzzz'
      }
    ]
  }
}
```

Once again we can display the labels to our users, and if they pick one, send back the value in the `_specialist` query parameter:

`/api/v1/products?_specialist=_cyyyy`

> We send the value, NOT the label. Again, you can send more than one by passing more than one `_specialist` query parameter. You'll get results that include at least one of the specialists.

### Adding counts for each distinct value

Want to show the user how many items are tagged `Free` as part of your filter interface? You can do that by using `distinct-counts` in place of `distinct`. Keep in mind that **the answer will still be in the `distinct` object**; however, each choice will now have a `count` property in addition to `label` and `value`.

Example request:

`/api/v1/products?distinct-counts=tags`

Example response:

```javascript
{
  results: [ ... pieces here ],
  distinct: {
    tags: [
      {
        label: 'Free',
        value: 'Free',
        count: 5
      },
      ... More tags here
    ]
  }
}
```


### Distinct values for more than one filter

Yes, this is supported. Just use comma-separated field names when passing `distinct` or `counts` in your URL.

For example, you might make this request:

`/api/v1/products?distinct=_specialist,tags`

In which case the `distinct` property of the response will have both `_specialist` and `tags` subproperties.

Make sure both `_specialist` and `tags` are configured as `safeDistinct`:

```javascript
// in lib/modules/products/index.js

'products': {
  name: 'product',
  extend: 'apostrophe-pieces',
  addFields: [
    {
      type: 'joinByOne',
      name: '_specialist'
    }
  ],
  restApi: {
    safeDistinct: [ '_specialist', 'tags' ]
  }
}
```

### Access as a logged-in user

If you are accessing the API as a user who can edit this piece type, you can use all cursor filters intended for web use, otherwise only the filters marked `safeFor: 'public'`.

## Retrieving one product

You can also retrieve one product via its `_id` property:

`/api/v1/products/cxxxxxxx`

The response is a single JSON object containing the product.

Even though you are fetching just one product, you can still invoke filters via the query string. If you are carrying out this request with the privileges of an admin user, you might want to add `?published=any` to gain access to an unpublished product.

## Inserting, updating and deleting products

These operations follow the usual REST patterns. But first, we need to talk about permissions.

## Invoking APIs when logged out

This is simple: if the user is not logged in, they will be able to `GET` public, published content, and that's all.

For many apps, **that's fine. You're using Apostrophe's admin bar to create the content anyway.**

Your content editors log into a site that's just for content creation, and your app users pull content from it via REST APIs. Great! **You're done here.**

But for those who need to create and manage content via REST too... read on!

## Invoking REST APIs as a logged-in user of your Apostrophe site

If you're building a React app or similar that is part of a webpage delivered by your Apostrophe site, and the right user is already logged into the site, then the APIs will automatically "see" the user and run with the right permissions. However, see the note that follows re: CSRF protection.

> If this doesn't sound relevant to your project, skip ahead to learn how to use API keys and bearer tokens instead. We've got your back, headless horseman.

## CSRF protection and logged-in users

**If an API request comes from an Apostrophe user who logged in conventionally via the website,** and not via the REST login APIs below, then Apostrophe will check for CSRF (Cross-Site Request Forgery) attacks.

If your API request is being sent by jQuery as provided by Apostrophe, you're good to go: Apostrophe automatically adds the necessary header.

If your API request is sent via `fetch` or another alternative to jQuery, you'll need to set the `X-XSRF-TOKEN` HTTP header to the current value of `window.apos.csrfCookieName`. This ensures the request didn't come from a sneaky form on a third-party website.

## Adding CORS config

It's also possible to add a specific CORS configuration for headless routes. Note that this does not secure your routes against the use of scripts, `curl`, etc. It only prevents well-behaved browsers like Chrome from making unwanted cross-site requests.
See available options [here](https://github.com/expressjs/cors#configuration-options)

```javascript
    'apostrophe-headless': {
      cors: {
        // CORS options
      }
    },
```

## Building apps without Apostrophe UI: bearer tokens and API keys

By default, the `POST`, `DELETE` and `PUT` APIs are available to logged-in users of the site. This is quite useful if you want to provide some editing features in a React or similar app that is part of your Apostrophe site.


CHANGELOG.md
## 2.12.1 2021-09-13
* Tests pass with latest `moog-require`. No code changes.

## 2.12.0 2021-05-20
* Adds ability to pass cors config to the module.

## 2.11.0 2020-12-02

* Added an `apostrophe-headless:beforeLogin` promise event. Thanks to Jose García of swiss4ward for the contribution.
* Eslint settings updated.

## 2.10.3 2020-11-04

* Fixes a typo in the README. Thanks to [Adrien Delannoy](https://github.com/Adrien-D) for the contribution. Updates `mocha` for a vulnerability warning, as well as ESLint.

## 2.10.2

* Fixes minor ESLint errors.

## 2.10.1

* Fixed bug in PATCH feature when using dot paths.

## 2.10.0

* Allows requests for pages to include `children=false` and `ancestors=false` query parameters that disable those respective properties on the response. Thanks to [Paul Grieselhuber](https://github.com/paulisloud) for the suggestion and original work on this feature.

## 2.9.5

* Updates the ESLint configuration to `eslint-config-apostrophe@3.10` and fixes linter errors.

## 2.9.4

* Make sure the slug follows the title when inserting a piece even if the slug is not explicitly POSTed.

## 2.9.3

* Fix issue introduced in apostrophe 2.102.0 where newly inserted pieces are unpublished, etc. You must update both this module and `apostrophe` to get the benefit of this fix.

## 2.9.2

* Fetches of a single piece (`GET /pieces/ID`) now respect excluded fields and `editPermissionRequired` in the same way that `GET /pieces` already did.

## 2.9.1

* The `csrf: exceptions` option to `apostrophe-express` now works properly in the presence of this module.

## 2.9.0

* Basic support for combining apostrophe-workflow with this module. GET requests now succeed as expected when the `_workflowLocale` query parameter is used to specify the desired locale. Otherwise, as before, you receive content for the default locale only. Note that you can only obtain live content, not draft content. POST, PUT and PATCH requests currently are not fully supported with workflow. Note that there is no issue if the doc type in question is excluded from workflow via `excludeTypes`. It is our intention to provide more complete support for headless workflow over time.

## 2.8.0

* Various tickets have been opened due to confusion around what happens if you use `includeFields` or `excludeFields` and, as a result, you do not include the following fields: `type`, `_id`, `tags`, `slug` and `docPermissions`. This can lead to the unexpected failure of joins, the unexpected absence of the `_url` property (or a bad value for that property), and the unexpected absence of the `_edit: true` property. This release fixes this issue by always including these fields in the MongoDB query, but excluding them after the fact in the returned array of results if they are present in `excludeFields`.
* The current `page` property is always included in the response when fetching pieces.
* eslint added to the tests, passes the apostrophe eslint config.

## 2.7.1

* The `PATCH` method works properly with `joinByOne` and `joinByArray`. You should send the appropriate `idField` or `idsField`. If these are not explicitly configured, the names map as follows: `_joinName` maps to `joinNameIdField` or `joinNameIdsField` (note there is no `_`), depending on whether it is a `joinByOne` or `joinByArray` field. Thanks to Giuseppe Monteleone for flagging the issue.
* In certain cases, a crash occurred when attempting to report a 500 error to the browser. Thanks to Giuseppe Monteleone for fixing the issue.

## 2.7.0

* `distinct` and `distinct-counts` query parameters added. You must also configure `safeDistinct`.

Thanks to Michelin for making this work possible via [Apostrophe Enterprise Support](https://apostrophecms.org/support/enterprise-support).

## 2.6.0

* `includeFields` and `excludeFields` now work properly for joins. Thanks to Anthony Tarlao.

## 2.5.0

* You may now specify just certain fields to be fetched with `includeFields` in your query string, or exclude certain fields with `excludeFields`. Thanks to `falkodev`.

## 2.4.0

* You may now exclude a field from the GET method of the API entirely by setting `api: false` in its schema field definition. You may also set `api: 'editPermissionRequired'` to restrict access to that field to those who can edit the doc in question. Thanks to Anthony Tarlao.
* If you would like to restrict GET access completely to those with edit permissions for the doc in question, you may now set the `getRequiresEditPermission` sub-option of `restApi` to `true`. Thanks again to Anthony Tarlao.

## 2.3.0

* Support for the `PATCH` method, which allows you to send just the fields you want to change, with support for simple array operators as well. Thanks to Paul Grieselhuber for his support.

## 2.2.0

* New `restApi.safeFilters` option (thanks to Marjan Georgiev), and documentation of the `restApi.maxPerPage` option.

## 2.1.2

* Documentation changes only. Clarified that areas must be present in the schema to be inserted or updated via the API.

## 2.1.1

* Fixed bug impacting the data provided by the `GET` route for pages when `all=1` is present. The data was incomplete due to missing query criteria.

## 2.1.0

* Support for API keys, as a lightweight alternative to bearer tokens for server-to-server communication. These should not be compiled into mobile apps, i.e. anywhere users might be able to obtain them by decompiling, etc.
* Support for pages, both reading and writing.
* Support for fragment rendering, and documentation on how to get fully rendered versions.

## 2.0.4

Documentation changes only. Gave some simple examples of query parameters that can be used to filter the results.

## 2.0.3

`apostrophe-headless` no longer has to be configured before modules it adds APIs to, and it is possible to add APIs to `apostrophe-images` or `apostrophe-files` if desired. Thanks to Stephen Walsh for pointing out the issue.

## 2.0.2

Documentation improvements. No code changes.

## 2.0.1


lib/modules/apostrophe-pieces-headless/index.js
var async = require('async');
var _ = require('lodash');

module.exports = {

  improve: 'apostrophe-pieces',

  construct: function(self, options) {

    self.addRestApiRoutes = function() {
      var restApi = self.apos.modules['apostrophe-headless'];
      if ((!options.restApi) || (options.restApi.enabled === false)) {
        return;
      }
      var baseEndpoint = restApi.endpoint;
      var endpoint = baseEndpoint + '/' + (options.restApi.name || self.__meta.name);

      // GET many
      self.apos.app.get(endpoint, function(req, res) {
        var cursor = self.findForRestApi(req);
        var result = {};
        return async.series([ distinct, countPieces, findPieces, renderPieces ], function(err) {
          if (err) {
            self.apos.utils.error(err);
            return res.status(500).send({ error: 'error' });
          }
          return res.send(result);
        });

        function distinct(callback) {
          var distinct = self.apos.launder.string(req.query.distinct).split(',');
          var counts = self.apos.launder.string(req.query['distinct-counts']).split(',');
          if (distinct[0] === '') {
            distinct = [];
          }
          if (counts[0] === '') {
            counts = [];
          }
          return async.eachSeries(_.uniq(distinct.concat(counts)), function(filter, callback) {
            if (!_.includes(self.options.restApi.safeDistinct || [], filter)) {
              return callback(null);
            }
            var counted = _.includes(counts, filter);
            var _cursor = cursor.clone();
            _cursor[filter](undefined);
            return _cursor.toChoices(filter, { counts: counted }, function(err, choices) {
              if (err) {
                return callback(err);
              }
              result.distinct = result.distinct || {};
              result.distinct[filter] = choices;
              return callback(null);
            });
          }, callback);
        }

        function countPieces(callback) {
          return cursor.toCount(function(err, count) {
            if (err) {
              return callback(err);
            }
            result.total = count;
            result.pages = cursor.get('totalPages');
            result.perPage = cursor.get('perPage');
            result.currentPage = cursor.get('page') || 1;
            return callback(null);
          });
        }

        function findPieces(callback) {
          return cursor.toArray(function(err, pieces) {
            if (err) {
              return callback(err);
            }
            // Attach `_url` and `_urls` properties
            self.apos.attachments.all(pieces, { annotate: true });
            pieces.forEach(function(piece) {
              self.restFilterFields(req, piece);
            });
            result.results = pieces;
            return callback(null);
          });
        }

        function renderPieces(callback) {
          return async.eachSeries(result.results, function(piece, callback) {
            var restApi = self.apos.modules['apostrophe-headless'];
            return restApi.apiRender(req, self, piece, 'piece', callback);
          }, callback);
        }

      });

      // GET one
      self.apos.app.get(endpoint + '/:id', function(req, res) {
        var id = self.apos.launder.id(req.params.id);
        if (!id) {
          return res.status(400).send({ error: 'bad request' });
        }
        var piece;
        return async.series([ find, render ], function(err) {
          if (err) {
            if (err === 'notfound') {
              return res.status(404).send({ error: 'notfound' });
            } else {
              self.apos.utils.error(err);
              return res.status(500).send({ error: 'error' });
            }
          }
          return res.send(piece);
        });
        function find(callback) {
          return self.findForRestApi(req).and({ _id: id }).toObject(function(err, _piece) {
            if (err) {
              return callback('error');
            }
            if (!_piece) {
              return callback('notfound');
            }
            piece = _piece;
            self.restFilterFields(req, piece);
            // Attach `_url` and `_urls` properties
            self.apos.attachments.all(piece, { annotate: true });
            return callback(null);
          });
        }
        function render(callback) {
          var restApi = self.apos.modules['apostrophe-headless'];
          return restApi.apiRender(req, self, piece, 'piece', callback);
        }
      });

      // POST one
      self.apos.app.post(endpoint, function(req, res) {
        var keys = Object.keys(req.body);
        if (_.includes(keys, 'title') && (!_.includes(keys, 'slug'))) {
          // Let the slug be inferred from the title properly
          keys.push('slug');
        }
        req.convertOnlyTheseFields = keys;
        return self.convertInsertAndRefresh(req, function(req, res, err, piece) {
          if (err) {
            return res.status(500).send({ error: 'error' });
          }
          return res.send(piece);
        });
      });

      // Update (PUT) one piece. The body must contain all of the properties
      // of the document as found in the schema otherwise they are
      // set blank. If this is not what you want, use the PATCH method.

      self.apos.app.put(endpoint + '/:id', function(req, res) {
        var id = self.apos.launder.id(req.params.id);
        return self.restPutOrPatch(req, id, 'put');
      });

      // PATCH one piece. Only touches properties present in the request;
      // also supports MongoDB-style `$push`, `$pullAll` and `$pullAllById` operators,
      // with a subset of the features found in MongoDB.
      //
      // PATCH operations are atomic with respect to other PATCH operations.
      //
      // PATCH operations may append to arrays using the
      // following syntax:
      //
      // `$push: { addresses: { street: '123 Wiggle Street' } }`
      //
      // The value given for `addresses` is appended to the existing
      // `addresses` schema field as a single element, even if it is itself
      // an array. This can be changed using the `$each` option:
      //
      // `$push: { addresses: { $each: [ { street: '123 Wiggle Street' }, { street: '101 Wacky Lane' } ] } }`
      //
      // Dot notation may be used to access arrays in subproperties with this
      // syntax.
      //
      // `$pullAll` may be used to remove all matching values present
      // for the array in question:
      //
      // `$pullAll: { addresses: { [ { street: '101 Wacky Lane', id: 'abcdef' } ] } }`
      //
      // If the array property in question is an
      // area's `items` property or an array schema field's value, it is more
      // convenient to remove array elements by their `id` or `_id` property:
      //
      // `$pullAllById`: 'abcdef'
      // `$pullAllById`: [ 'abcdef', 'qwerty' ]
      //
      // Note that this will match on either an `_id` property or an `-id` property.
      //
      // These operators can also be used to update the `idsField`
      // of a join. For instance, if a `joinByArray` field is named
      // `_people`, and `idsField` has not been set to the contrary,
      // you can append the `_id`s of additional people
      // to the `peopleIds` property using `$addToSet`.
      //
      // `patch` calls are guaranteed to be atomic with regard to
      // other `patch` operations. That is, if two `patch` operations
      // run concurrently updating different properties, all of the
      // property updates are guaranteed to make it through. If
      // two `patch` operations attempt to `$push` to the same
      // array, the first to begin will append its items first.

      self.apos.app.patch(endpoint + '/:id', function(req, res) {
        var id = self.apos.launder.id(req.params.id);
        return self.restPutOrPatch(req, id, 'patch');
      });

      self.restFilterFields = function(req, piece) {
        if (!piece._edit) {
          // Filter out editPermissions properties, where appropriate
          self.schema.forEach(function(field) {
            if (field.api === 'editPermissionRequired') {
              delete piece[field.name];
            }
          });
          // If you can't edit, it is none of your business who else can
          delete piece.docPermissions;
        }
        // To avoid situations that confuse developers, such as joins not
        // working or _url not populating, some fields like tags or type are
        // not excluded at the mongo level. Instead, delete them before
        // transmission
        _.each(req.excludeFields, function(field) {
          delete piece[field];
        });
      };

      // Implementation detail of the PUT and PATCH methods
      // (see above).
      //
      // Patch or put the given piece. The new data should be in `req.body` and
      // will be applied to the existing piece specified by `id` if
      // the permissions of `req` permit. If `action` is `patch`, only the
      // schema fields actually present in the `piece` object are touched,
      // otherwise all schema fields are touched, with absence treated
      // as an attempt to set an empty value for that property.

      self.restPutOrPatch = function(req, id, action) {
        if (action === 'patch') {
          return self.apos.locks.withLock('apostrophe-headless-' + id, body, respond);
        } else {
          return body(respond);
        }

        function respond(err) {
          if (err) {
            if (err === 'notfound') {
              return req.res.status(404).send({ error: err });
            } else if (err === 'invalid') {
              return req.res.status(400).send({ error: err });
            } else {
              return req.res.status(500).send({ error: 'error' });
            }
          }
          return req.res.send(req.piece);
        }

        function body(callback) {
          return self.findForEditing(req, { _id: id })
            .toObject(function(err, _piece) {
              if (err) {
                return callback(err);
              }
              if (!_piece) {
                return callback('notfound');
              }
              req.piece = _piece;
              if (action === 'patch') {
                var restApi = self.apos.modules['apostrophe-headless'];
                restApi.implementPatchOperators(_piece, req.body);
                req.restApiPatchSchema = restApi.subsetSchemaForPatch(self.schema, req.body);
              } else {
                req.convertOnlyTheseFields = Object.keys(req.body);
              }
              return self.convertUpdateAndRefresh(req, function(req, res, err, _piece) {
                return callback(err);
              });
            }
            );
        }

      };

      // DELETE one
      self.apos.app.delete(endpoint + '/:id', function(req, res) {
        var id = self.apos.launder.id(req.params.id);
        return async.series({
          before: function(callback) {
            return self.beforeTrash(req, id, callback);
          },
          trash: function(callback) {
            return self.trash(req, id, callback);
          },
          after: function(callback) {
            return self.afterTrash(req, id, callback);
          }
        }, function(err) {
          if (err) {
            return res.status(500).send({ error: 'error' });
          }
          return res.send({});
        });
      });

    };

    var superAllowedSchema = self.allowedSchema;
    self.allowedSchema = function(req) {
      var schema = superAllowedSchema(req);
      if (req.restApiPatchSchema) {
        schema = _.intersectionBy(schema, req.restApiPatchSchema, 'name');
      }
      return schema;
    };

    self.findForRestApi = function(req) {
      var which = 'public';
      var projection = {};
      var includeFromQuery = false;
      var joins = [];
      var joinsToExclude = [];

      if (req.query._workflowLocale) {
        // We don't use req.query.workflowLocale because that is caught by the
        // workflow middleware and would cause a redirect that is not useful here
        req.locale = req.query._workflowLocale;
      }

      if (req.query.includeFields) {
        // Always retrieve information necessary to annotate
        // what is editable and what is not, calculate _url, etc.
        projection.docPermissions = 1;
        projection.type = 1;
        projection.slug = 1;
        projection._id = 1;
        projection.tags = 1;
        var includeFields = self.apos.launder.string(req.query.includeFields).split(',');
        includeFields.forEach(function(field) {
          projection[field] = 1;
          includeFromQuery = true;
        });
      }

      if (self.apos.permissions.can(req, 'edit-' + self.name)) {
        which = 'manage';
      }

      self.schema.forEach(function(field) {
        if (field.api === false) {
          removeKey(field);
        }
        if (field.type.match(/join/)) {
          joins.push(field.name);
        } else if (field.schema) {
          field.schema.forEach(function(subField) {
            if (subField.type.match(/join/)) {
              joins.push(field.name + '.' + subField.name);
            }
          });
        }
      });

      if (!includeFromQuery && req.query.excludeFields) {
        req.excludeFields = self.apos.launder.string(req.query.excludeFields).split(',');
        req.excludeFields.forEach(function(field) {
          // Excluding these fields via mongodb has side effects on joins and _url that are
          // rarely anticipated by developers. We will delete them after the fetch
          if ((field !== 'type') && (field !== '_id') && (field !== 'tags') && (field !== 'slug')) {
            projection[field] = 0;
          }
        });
      }

      // add "exclude" fields only if "includeFields" fields are not required,
      // because Mongo cannot handle both at the same time
      function removeKey(field) {
        if (
          includeFromQuery ||
          Object.prototype.hasOwnProperty.call(projection, field.name)
        ) {
          delete projection[field.name];
          if (field.type.match(/join/)) {
            joinsToExclude.push(field.name);
          }
        } else {
          projection[field.name] = 0;
          if (field.type.match(/join/)) {
            if (field.relationship) {
              projection[field.relationshipsField] = 0;
            }
            if (field.idsField) {
              projection[field.idsField] = 0;
            }
            if (field.idField) {
              projection[field.idField] = 0;
            }
          }
        }

        return projection;
      }

      var joinsToInclude = _.difference(joins, joinsToExclude);
      var cursor = self.find(req, {})
        .projection(projection)
        .joins(joinsToInclude)
        .safeFilters((options.restApi.safeFilters || []).concat(options.restApi.safeDistinct || []))
        .queryToFilters(req.query, which);

      if (options.restApi.getRequiresEditPermission) {
        cursor.permission('edit');
      }

      var perPage = cursor.get('perPage');
      var maxPerPage = options.restApi.maxPerPage || 50;
      if ((!perPage) || (perPage > maxPerPage)) {
        cursor.perPage(maxPerPage);
      }

      var sort = req.query.sort;
      if (sort) {
        var sortBy = sort;
        var sortDirection = 1;
        if (sort.startsWith('-')) {
          sortBy = sort.substr(1);
          sortDirection = -1;
        }
        var sortExpression = {};
        sortExpression[sortBy] = sortDirection;
        cursor.sort(sortExpression);
      }
      return cursor;
    };

    self.modulesReady = function() {
      var restApi = self.apos.modules['apostrophe-headless'];
      self.addRestApiRoutes();
      restApi.registerModule(self);
    };
  }
};
